sdk/core/azure-core/inc/http/http.hpp

```diff
@@ -214,21 +238,33 @@ class Response
 
 public:
   Response(uint16_t statusCode, std::string const& reasonPhrase)
-      : Response(statusCode, reasonPhrase, http::BodyBuffer::null, http::BodyStream::null)
+      : Response(statusCode, reasonPhrase, std::vector<uint8_t>(), http::BodyStream::null)
   {
   }
 
+  // default constructor for bulding from base response
+  Response() : Response(0, "", std::vector<uint8_t>(), http::BodyStream::null) {}
+
   // Methods used to build HTTP response
+  void setVersion(uint16_t mayorVersion, uint16_t minorVersion)
+  {
+    m_httpVersion = http::HttpVersion(mayorVersion, minorVersion);
+  }
+  void setStatusCode(uint16_t statusCode) { m_statusCode = statusCode; }
+  void setReasonPhrase(std::string reasonPhrase) { m_reasonPhrase = reasonPhrase; }
   void addHeader(std::string const& name, std::string const& value);
-  void setBody(BodyBuffer* bodyBuffer);
   void setBody(BodyStream* bodyStream);
```

Comment:
This also will probably not exist and should be passed via the ctor.
In general, it is best to avoid setters whenever possible. The reason is because they can be called at any time and nothing prevents this. So, your code could create a Response and set a body. Then some other code (that you may not have control over) may call a setter and change the object's state. Then some other code (that you or may not have control over) may access the object and the original value set has now been overwritten by other code. Bugs may now occur and you'll never be able to find them because the code that modified the object's state is not on the thread stack any more and so you can't find out what code mutated the object and how that code ever executed. In addition, making objects immutable also makes them thread safe and so this gives even more incentive to follow this pattern as it prevents very, very hard bugs to find and fix.

So, it is best to try really hard to construct objects that then become immutable. 
In this class, AddHeader and AppendBody (which should both start with uppercase letters based on our guidelines decision last week), do not follow this rule of thumb as any code could call AddHeader or AppendBody at any time. This is unfortunate but we can probably improve this (by removing these methods) at a later time.
Regardless of whether we remove them or not, reducing the number of setters to a minimum is always something to really strive for.